src/array_like.rs

```diff
@@ -0,0 +1,81 @@
+//! TODO
+
+use crate::{get_array_module, ndarray::Dimension, Element, PyReadonlyArray};
+use ndarray::{Ix1, Ix2, Ix3, Ix4, Ix5, Ix6, IxDyn};
+use pyo3::{types::PyDict, FromPyObject, PyAny, PyResult};
+use std::ops::Deref;
+
+/// TODO
+#[derive(Debug)]
+pub struct PyArrayLike<'py, T, D>(PyReadonlyArray<'py, T, D>)
+where
+    T: Element,
+    D: Dimension;
+
+impl<'py, T, D> Deref for PyArrayLike<'py, T, D>
+where
+    T: Element,
+    D: Dimension,
+{
+    type Target = PyReadonlyArray<'py, T, D>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<'py, T, D> FromPyObject<'py> for PyArrayLike<'py, T, D>
+where
+    T: Element,
+    D: Dimension,
+    Vec<T>: FromPyObject<'py>,
+{
+    fn extract(ob: &'py PyAny) -> PyResult<Self> {
+        if let Ok(array) = ob.extract::<PyReadonlyArray<T, D>>() {
+            return Ok(PyArrayLike(array));
+        }
+
+        let py = ob.py();
+
+        // The following does not work because of dimension type conflicts.
+        // Maybe we should implement `into_dimensionality` for PyArray?
```

Comment:
I think you can do this conversion via `Array1::from_vec`, then `into_dimensionality` and then `.into_pyarray` which should all be zero cost.